From 607804ef8c7ed57a1e1916f27b9e41fde9e37960 Mon Sep 17 00:00:00 2001
From: Amelie Delaunay <amelie.delaunay@foss.st.com>
Date: Tue, 25 Nov 2025 09:38:30 +0100
Subject: [PATCH] Revert "serial: stm32: retrieve data stuck in DMA FIFO"

This reverts commit 83771e186d2abeccc38b0a5e6841b068499622ca.

This feature was introduced with STM32MP21 DMA3, but some bytes can be
lost. Another implementation is required. Pending that, do not force
draining of DMA flight data. Instead, avoid DMA packing so bytes are not
stored in the DMA FIFO before being pushed to the destination buffer.

Change-Id: Ie61b8e1e46f422b40ac4f9dcd04e54b35ba24a19
Signed-off-by: Amelie Delaunay <amelie.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/512806
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/tty/serial/stm32-usart.c | 26 +++++---------------------
 1 file changed, 5 insertions(+), 21 deletions(-)

diff --git a/drivers/tty/serial/stm32-usart.c b/drivers/tty/serial/stm32-usart.c
index 5dae07643a78..00e0d8d46df0 100644
--- a/drivers/tty/serial/stm32-usart.c
+++ b/drivers/tty/serial/stm32-usart.c
@@ -113,7 +113,6 @@ static struct stm32_usart_info __maybe_unused stm32h7_info = {
 static void stm32_usart_stop_tx(struct uart_port *port);
 static void stm32_usart_transmit_chars(struct uart_port *port);
 static void __maybe_unused stm32_usart_console_putchar(struct uart_port *port, unsigned char ch);
-static int stm32_usart_rx_dma_start_or_resume(struct uart_port *port);
 
 static inline struct stm32_port *to_stm32_port(struct uart_port *port)
 {
@@ -652,33 +651,18 @@ static unsigned int stm32_usart_receive_chars(struct uart_port *port, bool force
 {
 	struct stm32_port *stm32_port = to_stm32_port(port);
 	const struct stm32_usart_offsets *ofs = &stm32_port->info->ofs;
-	struct dma_tx_state *state;
 	enum dma_status rx_dma_status;
 	u32 sr;
 	unsigned int size = 0;
 
 	if (stm32_usart_rx_dma_started(stm32_port) || force_dma_flush) {
-		state = &stm32_port->rx_dma_state;
 		rx_dma_status = dmaengine_tx_status(stm32_port->rx_ch,
-						    stm32_port->rx_ch->cookie, state);
+						    stm32_port->rx_ch->cookie,
+						    &stm32_port->rx_dma_state);
 		if (rx_dma_status == DMA_IN_PROGRESS ||
 		    rx_dma_status == DMA_PAUSED) {
-			if (force_dma_flush && state->in_flight_bytes) {
-				/* Disable RX DMA to force in flight data is drained */
-				stm32_usart_rx_dma_terminate(stm32_port);
-				state->residue -= state->in_flight_bytes;
-				/* Empty DMA buffer */
-				size = stm32_usart_receive_chars_dma(port);
-				if (rx_dma_status == DMA_IN_PROGRESS) {
-					/* If restarting DMA fails, fall back to interrupt mode */
-					if (stm32_usart_rx_dma_start_or_resume(port))
-						size = stm32_usart_receive_chars_pio(port);
-				}
-			} else {
-				/* Empty DMA buffer */
-				size = stm32_usart_receive_chars_dma(port);
-			}
-
+			/* Empty DMA buffer */
+			size = stm32_usart_receive_chars_dma(port);
 			sr = readl_relaxed(port->membase + ofs->isr);
 			if (sr & USART_SR_ERR_MASK) {
 				/* Disable DMA request line */
@@ -1172,7 +1156,7 @@ static irqreturn_t stm32_usart_interrupt(int irq, void *ptr)
 	/* Receiver timeout irq for DMA RX */
 	if (stm32_usart_rx_dma_started(stm32_port) && !stm32_port->throttled) {
 		spin_lock(&port->lock);
-		size = stm32_usart_receive_chars(port, (sr & USART_SR_RTOF));
+		size = stm32_usart_receive_chars(port, false);
 		if (size)
 			tty_flip_buffer_push(tport);
 		uart_unlock_and_check_sysrq(port);
-- 
2.53.0

